src/HealthChecks/HealthChecks/src/HealthCheckPublisherHostedService.cs

```diff
@@ -78,16 +102,37 @@ public Task StopAsync(CancellationToken cancellationToken = default)
             return Task.CompletedTask;
         }
 
-        _timer?.Dispose();
-        _timer = null;
+        if (_timersByOptions != null)
+        {
+            foreach (var timer in _timersByOptions.Values)
+            {
+                timer.Dispose();
+            }
+
+            _timersByOptions = null;
+        }
 
         return Task.CompletedTask;
     }
 
-    // Yes, async void. We need to be async. We need to be void. We handle the exceptions in RunAsync
-    private async void Timer_Tick(object? state)
+    private Dictionary<(TimeSpan Delay, TimeSpan Period), Timer> CreateTimers(IReadOnlyDictionary<(TimeSpan Delay, TimeSpan Period), List<HealthCheckRegistration>> healthChecksByOptions)
+    {
+        return healthChecksByOptions.Select(m => CreateTimer(m.Key)).ToDictionary(kv => kv.Key, kv => kv.Value);
+    }
```

Comment:
This saves a lot of intermediate steps. Nothing is indexing into this Timer Dictionary. The caller and CreateTimer will have to be updated for this to work.

```suggestion
    private Timer[] CreateTimers(IEnumerable<IGrouping<(TimeSpan Delay, TimeSpan Period), HealthCheckRegistration>> groupedRegistrations)
    {
        return groupedRegistrations.Select(g => CreateTimer(g.Key.Delay, g.Key.Period)).ToArray();
    }
```

Reply:
Refactored to LINQ to foreach + List<Timer>